### ISSUE TITLE
ITransaction.CommitAsync() takes a long time to complete

### ISSUE THREAD

**Issue by MikeJohnson1337** [Azure/service-fabric-issues#253]
Hi,

I am having a problem where ITransaction.CommitAsync() takes a long time to complete (sometimes over 24 hours), or never does. See http://stackoverflow.com/questions/38737409/what-can-cause-itransaction-commitasync-calls-to-take-a-really-long-time-24h for more details.

I have verified this with breakpoints.

I am using Service Remoting to enqueue items, as such:

```
public async Task EnqueueAsync(long id)
{
	TraceInformation("Queueing {0}...", id);

	IReliableQueue<long> exampleQueue = await StateManager.GetOrAddAsync<IReliableQueue<long>>("exampleQueue");

	using (ITransaction transaction = StateManager.CreateTransaction())
	{
		await exampleQueue.EnqueueAsync(transaction, id);

		await transaction.CommitAsync();
	}
}
```

The RunAsync method is as such (I have put in comments where it hangs):

```
protected override async Task RunAsync(CancellationToken cancellationToken)
{
	IReliableQueue<long> exampleQueue = await StateManager.GetOrAddAsync<IReliableQueue<long>>("exampleQueue");

	while (!cancellationToken.IsCancellationRequested)
	{
		using (ITransaction transaction = StateManager.CreateTransaction())
		{
			ConditionalValue<long> queueItem =
				await exampleQueue.TryDequeueAsync(transaction, TimeSpan.FromSeconds(4), cancellationToken);

			if (queueItem.HasValue)
			{
				ProcessQueueItem(queueItem.Value);
			}

			// If an exception is thrown before calling CommitAsync, the transaction aborts,
			// all changes are discarded, and nothing is saved to the secondary replicas.
			await transaction.CommitAsync(); // <-- Sits here
		}

		await Task.Delay(TimeSpan.FromSeconds(1), cancellationToken); // <-- Never reaches here
	}
}
```

Thanks,
Mike

**Comment by MikeJohnson1337** [comment]
There appears to be a time limit on the transaction. I verified this by running half of my 'work' code. Is this true?

**Comment by MikeJohnson1337** [comment]
My last comment was wrong. It appears that you cannot use Windows Forms Controls. Is there any workaround for this?

**Comment by MikeJohnson1337** [comment]
For anyone interested, you need to call Dispose() on the control before calling CommitAsync(), otherwise it will sit there forever.

**Comment by johnbindel** [comment]
I'm guessing that you are using Service Fabric Remoting to get a Windows Form to talk to a service in your Service Fabric cluster. If that's the case, the issue is related to your use of async in the UI.

**Preventing Async UI Deadlock**
If you are calling async methods from a winforms app, you will need to ensure that the async code doesn't capture the UI context, which would cause a deadlock.

To do this, you'll need to use the following syntax for the awaits within your library code:
`await xyz().ConfigureAwait(false);`
This will prevent the deadlock by allowing a non-UI context to resume execution after the awaited call completes.

For more info, I would recommend looking at Stephen Cleary's helpful blog posts, such as: https://blog.stephencleary.com/2012/07/dont-block-on-async-code.html

**Service Fabric Remoting**
If you are using Service Fabric Remoting to make the call, then the ConfigureAwait(false) should be added onto that call. It will not need any changes inside your Service Fabric service code as that is hosted in a different process (perhaps even in the cloud) and does not have a UI context to deadlock with.

Hope this helps!

**Comment by masnider** [comment]
hi @MikeJohnson1337 - were you able to look at @johnbindel 's suggestions? It's correct that most of the time we have seen this type of behavior it's been some synchronous waiting on an async task or thread. 

If you could get back to us and let us know if you've gotten things working, or let us know if you're still having issues, that'd be great. If we don't hear back from you in a bit we'll close this thread.

**Comment by masnider** [comment]
I'm closing this now. let us know if you still need help by reopening this or creating a new issue. Thanks!

**Comment by MikeJohnson1337** [comment]
@johnbindel's suggestions helped. Thanks.
